kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java

```diff
@@ -0,0 +1,158 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.stream.annotation.Bindings;
+import org.springframework.cloud.stream.annotation.EnableBinding;
+import org.springframework.cloud.stream.messaging.Source;
+import org.springframework.context.annotation.Bean;
+import org.springframework.integration.dsl.IntegrationFlow;
+import org.springframework.integration.dsl.IntegrationFlows;
+import org.springframework.integration.dsl.kafka.Kafka;
+import org.springframework.integration.kafka.core.*;
+import org.springframework.integration.kafka.listener.KafkaMessageListenerContainer;
+import org.springframework.integration.kafka.listener.KafkaNativeOffsetManager;
+import org.springframework.integration.kafka.listener.OffsetManager;
+import org.springframework.integration.kafka.support.ZookeeperConnect;
+import org.springframework.retry.support.RetryTemplate;
+import org.springframework.util.Assert;
+import org.springframework.util.ObjectUtils;
+import org.springframework.util.StringUtils;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Source module that receives data from Kafka.
+ *
+ * @author Soby Chacko
+ */
+@EnableBinding(Source.class)
+@EnableConfigurationProperties(KafkaConfigurationProperties.class)
+public class KafkaSource {
+
+	@Autowired
+	@Bindings(KafkaSource.class)
+	private Source source;
+
+	@Autowired
+	private KafkaConfigurationProperties properties;
+
+	@Bean
+	public KafkaMessageListenerContainer container() {
+
+		Assert.isTrue(!ObjectUtils.isEmpty(properties.getTopics()) || !ObjectUtils.isEmpty(properties.getPartitions()),
+				"Either a list of topics OR a topic=<comma separated partitions> must be provided");
+
+		Configuration zookeeperConfiguration = new ZookeeperConfiguration(zookeeperConnect());
+		ConnectionFactory kafkaConnectionFactory = new DefaultConnectionFactory(zookeeperConfiguration);
+
+		Partition[] partitions = getPartitions();
+
+		KafkaMessageListenerContainer container = ObjectUtils.isEmpty(partitions) ?
+				new KafkaMessageListenerContainer(kafkaConnectionFactory, properties.getTopics()) :
+				new KafkaMessageListenerContainer(kafkaConnectionFactory, partitions);
+
+		container.setOffsetManager(kafkaNativeOffsetManager(kafkaConnectionFactory));
+
+		return container;
+	}
+
+	@Bean
+	public ZookeeperConnect zookeeperConnect() {
+		ZookeeperConnect zookeeperConnect = new ZookeeperConnect();
+		zookeeperConnect.setZkConnect(properties.getZkConnect());
+		zookeeperConnect.setZkConnectionTimeout(properties.getZkConnectionTimeout());
+		zookeeperConnect.setZkSessionTimeout(properties.getZkSessionTimeout());
+		zookeeperConnect.setZkSyncTime(properties.getZkSyncTime());
+		return zookeeperConnect;
+	}
+
+	@Bean
+	public OffsetManager kafkaNativeOffsetManager(ConnectionFactory kafkaConnectionFactory) {
```

Comment:
I don't see the `ConnectionFactory` `@Bean`.
Who populates it?
What have I missed? Why then `container()` bean doesn't use the same injection technique?